pkg/blobstore/spanner_blob_access.go

```diff
@@ -0,0 +1,975 @@
+		log.Printf("GET: scheduling touch reftime for key %s reftime %s", key, s.ReferenceTime)
+		ba.refChan <- key
+	}
+	return b
+}
+
+func (ba *spannerBlobAccess) Put(ctx context.Context, digest digest.Digest, b buffer.Buffer) error {
+	//log.Printf("SpannerBlobAccess type %#+v PUT digest %s", ba.storageType, digest)
+	if err := util.StatusFromContext(ctx); err != nil {
+		b.Discard()
+		return err
+	}
+
+	// If we're bigger than 10MB, we have to offload to GCS
+	size, err := b.GetSizeBytes()
+	if err != nil {
+		log.Printf("Put Blob %s: can't get size: %v", digest, err)
+		b.Discard()
+		return err
+	}
+
+	key := ba.digestToKey(digest)
+	//log.Printf("SpannerBlobAccess PUT key is %s, size %d", key, size)
+
+	var inlineData []byte = nil
+	if ba.storageType == pb.StorageType_CASTORE {
+		statedSize, err := ba.keyToBlobSize(key)
+		if err != nil {
+			spannerMalformedKeyCount.Inc()
+			log.Printf("Put Blob: invalid digest size, key %s: %v", key, err)
+			b.Discard()
+			return err
+		}
+		if statedSize != size {
+			log.Printf("Put Blob: MISMATCH between buffer size (%d) and digest size (%d)", size, statedSize)
+			b.Discard()
+			return fmt.Errorf("Invalid contents digest %s size %d", digest, size)
+		}
+	}
+	now := time.Now().UTC()
+	if size > maxSize {
+		obj := ba.gcsBucket.Object(key)
+		w := obj.NewWriter(ctx)
+		start := time.Now()
+		var err error
+		if _, err = io.Copy(w, b.ToReader()); err == nil {
+			if err = w.Close(); err != nil {
+				log.Printf("Blob %s can't be copied to GCS, close failed: %v", digest, err)
+			}
+		} else {
+			log.Printf("Blob %s can't be copied to GCS, write failed: %v", digest, err)
+		}
+		backendOperationsDurationSeconds.WithLabelValues(BEMap[ba.storageType], BE_GCS, BE_PUT).Observe(time.Now().Sub(start).Seconds())
+		if err != nil {
+			if errors.Is(err, context.Canceled) || status.Code(err) == codes.Canceled {
+				gcsPutFailedContextCanceledCount.Inc()
+			} else if errors.Is(err, context.DeadlineExceeded) || status.Code(err) == codes.DeadlineExceeded {
+				gcsPutFailedDeadlineExceededCount.Inc()
+			} else {
+				gcsPutFailedOtherCount.Inc()
+			}
+			return err
+		}
+		inlineData = nil
+		ba.touchGCSObject(context.Background(), key, now)
+	} else {
+		inlineData, err = b.ToByteSlice(int(maxSize))
+		if err != nil {
+			log.Printf("Blob %s can't be copied to Spanner: %v", digest, err)
+			return err
+		}
+		if len(inlineData) == 0 {
+			log.Printf("WARNING: ByteSlice size is 0, expected %d, key %s", size, key)
+		}
+	}
+
+	// Now insert into the spanners no matter what!
+	rec := spannerRecord{
+		Key:           key,
+		InlineData:    inlineData,
+		ReferenceTime: now,
+	}
+	insertMut, err := spanner.ReplaceStruct(tableName, rec)
+	if err != nil {
+		log.Printf("Can't create mutation for Blob %s: %v", digest, err)
+		return err
+	}
+
+	start := time.Now()
+	_, err = ba.spannerClient.Apply(context.Background(), []*spanner.Mutation{insertMut})
```

Comment:
Why use `context.Background()` here?

Reply:
This one should be ctx.  I use context.Background when I want to perform an action, but have it unaffected if the context gets cancelled.